File: matiasmasso/Mat.NET/DTO/DTOEci.cs
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOEci
    {
        public static string NumeroDeComanda(DTODelivery oDelivery)
        {
            string retval = "";
            switch (oDelivery.Cod)
            {
                case DTOPurchaseOrder.Codis.client:
                    {
                        DTOPurchaseOrder oOrder = oDelivery.Items.First().PurchaseOrderItem.PurchaseOrder;
                        retval = NumeroDeComanda(oOrder);
                        break;
                    }
            }
            return retval;
        }

        public static string NumeroDeComanda(DTOPurchaseOrder oOrder)
        {
            string retval = "";
            if (!string.IsNullOrEmpty(oOrder.Concept))
            {
                string sPdd = oOrder.Concept;
                int i;
                for (i = 0; i <= sPdd.Length - 1; i++)
                {
                    if (!"0123456789".Contains(sPdd.Substring(i, 1)))
                        break;
                }
                if (i > 0)
                    retval = sPdd.Substring(0, i);
            }
            return retval;
        }

        public static string GetDepartamentoFromFirstAlbPdd(List<DTODelivery> oDeliveries)
        {
            string retval = "";
            // /dep.053/
            foreach (DTODelivery oDelivery in oDeliveries)
            {
                foreach (DTODeliveryItem oItem in oDelivery.Items)
                {
                    if (oItem.PurchaseOrderItem != null)
                    {
                        DTOPurchaseOrder oOrder = oItem.PurchaseOrderItem.PurchaseOrder;
                        string src = oOrder.Concept;
                        int iStart = src.IndexOf("/dep.");
                        if (iStart >= 0)
                        {
                            int iEnd = src.IndexOf("/", iStart + 5);
                            if (iEnd > iStart)
                            {
                                string sSegment = src.Substring(iStart, iEnd - iStart);
                                retval = sSegment.Replace("dep.", "").Replace("/", "");
                                break;
                            }
                        }
                    }
                }
                if (!string.IsNullOrEmpty(retval))
                    break;
            }
            return retval;
        }

        public static void GetDetailsFromPdc(DTOPurchaseOrder oPurchaseOrder, ref string sPedido, ref string sCentro, ref string sDepartamento, ref string sNumProveedor)
        {
            // 30560101/ctro.0050/dep.053/prov.01-030825
            string[] sPdd = oPurchaseOrder.Concept.Split('/');
            if (sPdd.Length == 4)
            {
                sPedido = sPdd[0];
                if (sPdd[1].Length >= 5)
                    sCentro = sPdd[1].Substring(5);
                sDepartamento = sPdd[2].Replace("dep.", "");
                sNumProveedor = sPdd[3].Replace("prov.", "");
            }
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOEci.vb
Public Class DTOEci
    Shared Function NumeroDeComanda(ByVal oDelivery As DTODelivery) As String
        Dim retval As String = ""
        Select Case oDelivery.Cod
            Case DTOPurchaseOrder.Codis.Client
                Dim oOrder As DTOPurchaseOrder = oDelivery.Items.First.PurchaseOrderItem.PurchaseOrder
                retval = NumeroDeComanda(oOrder)
        End Select
        Return retval
    End Function

    Shared Function NumeroDeComanda(ByVal oOrder As DTOPurchaseOrder) As String
        Dim retval As String = ""
        If oOrder.Concept > "" Then
            Dim sPdd As String = oOrder.Concept
            Dim i As Integer
            For i = 0 To sPdd.Length - 1
                If Not "0123456789".Contains(sPdd.Substring(i, 1)) Then Exit For
            Next
            If i > 0 Then retval = sPdd.Substring(0, i)
        End If
        Return retval
    End Function

    Shared Function GetDepartamentoFromFirstAlbPdd(oDeliveries As List(Of DTODelivery)) As String
        Dim retval As String = ""
        '/dep.053/
        For Each oDelivery As DTODelivery In oDeliveries
            For Each oItem As DTODeliveryItem In oDelivery.Items
                If oItem.PurchaseOrderItem IsNot Nothing Then
                    Dim oOrder As DTOPurchaseOrder = oItem.PurchaseOrderItem.PurchaseOrder
                    Dim src As String = oOrder.Concept
                    Dim iStart As Integer = src.IndexOf("/dep.")
                    If iStart >= 0 Then
                        Dim iEnd As Integer = src.IndexOf("/", iStart + 5)
                        If iEnd > iStart Then
                            Dim sSegment As String = src.Substring(iStart, iEnd - iStart)
                            retval = sSegment.Replace("dep.", "").Replace("/", "")
                            Exit For
                        End If
                    End If
                End If
            Next
            If retval > "" Then Exit For
        Next
        Return retval
    End Function

    Shared Sub GetDetailsFromPdc(ByVal oPurchaseOrder As DTOPurchaseOrder, ByRef sPedido As String, ByRef sCentro As String, ByRef sDepartamento As String, Optional ByRef sNumProveedor As String = "")
        '30560101/ctro.0050/dep.053/prov.01-030825
        Dim sPdd() As String = oPurchaseOrder.Concept.Split("/")
        If sPdd.Length = 4 Then
            sPedido = sPdd(0)
            If sPdd(1).Length >= 5 Then
                sCentro = sPdd(1).Substring(5)
            End If
            sDepartamento = sPdd(2).Replace("dep.", "")
            sNumProveedor = sPdd(3).Replace("prov.", "")
        End If
    End Sub
End Class
